# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/CanvasTranslator.cpp
# Commit: 770a728fe2b7
# Full Hash: 770a728fe2b706a3035710979f4ef8e4e29f89c4
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2023-12-13 21:42:49
# Description:
#   Bug 1869659: Check for Failed in CanvasTranslator::AddBuffer and SetDataSurfaceBuffer. r=aosmond
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D196287
# ==============================================================================

diff -r 381eaa0ce60f -r 770a728fe2b7 gfx/layers/ipc/CanvasTranslator.cpp
--- a/gfx/layers/ipc/CanvasTranslator.cpp	Wed Dec 13 14:28:12 2023 +0000
+++ b/gfx/layers/ipc/CanvasTranslator.cpp	Wed Dec 13 14:33:25 2023 +0000
@@ -219,6 +219,11 @@
 void CanvasTranslator::AddBuffer(ipc::SharedMemoryBasic::Handle&& aBufferHandle,
                                  size_t aBufferSize) {
   MOZ_ASSERT(IsInTaskQueue());
+  if (mHeader->readerState == State::Failed) {
+    // We failed before we got to the pause event.
+    return;
+  }
+
   MOZ_RELEASE_ASSERT(mHeader->readerState == State::Paused);
   MOZ_ASSERT(mDefaultBufferSize != 0);
 
@@ -258,6 +263,11 @@
 void CanvasTranslator::SetDataSurfaceBuffer(
     ipc::SharedMemoryBasic::Handle&& aBufferHandle, size_t aBufferSize) {
   MOZ_ASSERT(IsInTaskQueue());
+  if (mHeader->readerState == State::Failed) {
+    // We failed before we got to the pause event.
+    return;
+  }
+
   MOZ_RELEASE_ASSERT(mHeader->readerState == State::Paused);
 
   if (!CreateAndMapShmem(mDataSurfaceShmem, std::move(aBufferHandle),
@@ -504,6 +514,7 @@
               gfxCriticalNote << "Failed to read event type: "
                               << recordedEvent->GetType();
             }
+            mHeader->readerState = State::Failed;
             return false;
           }
 
@@ -523,6 +534,7 @@
       } else {
         gfxCriticalNote << "Failed to play canvas event type: " << eventType;
       }
+      mHeader->readerState = State::Failed;
     }
 
     mHeader->processedCount++;
